Enhances exists queries to reduce need for `_field_names`

colings86 opened the pull request:
Before this change we wrote the name all the fields in a document to a `_field_names` field and then implemented exists queries as a term query on this field. The problem with this approach is that it bloats the index and also affects indexing performance.

This change adds a new method `existsQuery()` to `MappedFieldType` which is implemented by each sub-class. For most field types if doc values are available a `DocValuesFieldExistsQuery` is used, falling back to using `_field_names` if doc values are disabled. Note that only fields where no doc values are available are written to `_field_names`.

Closes #26770

colings86 committed:
Enhances exists queries to reduce need for `_field_names`

Before this change we wrote the name all the fields in a document to a `_field_names` field and then implemented exists queries as a term query on this field. The problem with this approach is that it bloats the index and also affects indexing performance.

This change adds a new method `existsQuery()` to `MappedFieldType` which is implemented by each sub-class. For most field types if doc values are available a `DocValuesFieldExistsQuery` is used, falling back to using `_field_names` if doc values are disabled. Note that only fields where no doc values are available are written to `_field_names`.

Closes #26770

jpountz commented on core/src/main/java/org/elasticsearch/index/query/ExistsQueryBuilder.java:
```diff
@@ -143,18 +149,49 @@ public static Query newFilter(QueryShardContext context, String fieldPattern) {
+            }
+        }
+        return new ConstantScoreQuery(booleanQuery.build());
+    }
```
Now that I think more about it, I think this impl has potential problems with multiple types, since getObjectMapper will return the mapper of one of the types, meaning it might miss sub fields if they only exist in another type (those mapper classes are per-type on the contrary to the field type classes). So maybe we should resolve fields that match field + ".*" and build a bool query out of it (and no need for recursion since I would expect the glob to match fields of sub objects too)

colings86 replied:
fixed

jpountz commented on core/src/main/java/org/elasticsearch/index/mapper/TextFieldMapper.java:
```diff
@@ -333,6 +333,7 @@ protected void parseCreateField(ParseContext context, List<IndexableField> field
             Field field = new Field(fieldType().name(), value, fieldType());
             fields.add(field);
         }
+        createFieldNamesField(context, fields);
```
maybe use NormFieldExistsQuery since we are on a 7.1 snapshot now?

colings86 replied:
I'll do this as a followup PR as it will require the Lucene 7.1 snapshot to be updated too

jpountz replied:
+1

jpountz reviewed (commented):
Thanks for working on it.

One concern I have is that the impl for existsQuery is shared in a super class while the logic for adding mappers is in concrete impls. I think it makes it very easy to inherit the default impl for the exists query and forget to add a _field_names field at index time if necessary?

colings86 commented on client/rest/src/main/java/org/elasticsearch/client/RestClientBuilder.java:
```diff
@@ -39,7 +39,7 @@
 public final class RestClientBuilder {
     public static final int DEFAULT_CONNECT_TIMEOUT_MILLIS = 1000;
-    public static final int DEFAULT_SOCKET_TIMEOUT_MILLIS = 30000;
+    public static final int DEFAULT_SOCKET_TIMEOUT_MILLIS = 3000000;
```
haha good spot, this was so that the rest client didn't time out when I was debugging rest tests, shouldn't have been committed.

colings86 commented on core/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java:
```diff
@@ -285,6 +287,18 @@ public Mapper parse(ParseContext context) throws IOException {
+                .metadataMapper(FieldNamesFieldMapper.class).fieldType();
+        if (fieldNamesFieldType != null && fieldNamesFieldType.isEnabled()) {
+            for (String fieldName : FieldNamesFieldMapper.extractFieldNames(fieldType().name())) {
+                if (fieldType().indexOptions() != IndexOptions.NONE || fieldType().stored()) {
```
@jpountz and I discussed this and we agreed that for this PR we will keep the behaviour that exists queries will only work for fields that are indexed, stored, or have doc values (or a combination of those) so this PR can be merged into 7.0 and 6.x. I'll make a followup PR for 7.0 only that changes things to allow exists to work for any field that is mapped.

jpountz replied:
Hmm but then that if statement is still buggy since it should also check for `hasDocValues`? Actually I'd be in favour of removing it entirely since you seem to have that logic on the call sites already anyway (only adding this field if there are no doc values)?

colings86 replied:
I've moved this logic outside of this method, the callers now check to ensure if field names field should be created

colings86 commented on core/src/main/java/org/elasticsearch/index/mapper/BinaryFieldMapper.java:
```diff
@@ -20,6 +20,7 @@
+
 import org.apache.lucene.document.Field;
 import org.apache.lucene.index.IndexOptions;
 import org.apache.lucene.index.IndexableField;
```
see https://github.com/elastic/elasticsearch/pull/26930#discussion_r143664446

jpountz replied:
Then shouldn't it create field names fields when it is stored or has doc values?

colings86 replied:
You are right, I've made changes to support this

colings86 commented on core/src/main/java/org/elasticsearch/index/mapper/SourceFieldMapper.java:
```diff
@@ -163,6 +162,11 @@ public String typeName() {
         }
 
         @Override
+        public Query existsQuery(QueryShardContext context) {
```
I threw an exception here to mirror the behaviour of `termQuery()`. I assumed we did this in term query to discourage people from performing queries on `_source` and assumed we would want to do this here too for the same reason?

jpountz replied:
+1

colings86 committed:
Addresses review comments

colings86 commented on core/src/main/java/org/elasticsearch/index/query/ExistsQueryBuilder.java:
```diff
@@ -143,18 +149,49 @@ public static Query newFilter(QueryShardContext context, String fieldPattern) {
+            } else {
+                Query filter = fieldType.existsQuery(context);
+                boolFilterBuilder.add(filter, BooleanClause.Occur.SHOULD);
+            }
```
fixed

colings86 commented on core/src/main/java/org/elasticsearch/index/mapper/FieldNamesFieldMapper.java:
```diff
@@ -206,7 +206,6 @@ public void preParse(ParseContext context) throws IOException {
 
     @Override
     public void postParse(ParseContext context) throws IOException {
-        super.parse(context);
```
done

colings86 commented:
@jpountz I pushed a commit that addresses your comments. Could you take another look?

jpountz reviewed (commented):
I left some more comments. I think we also need to fix the `_field_names` docs to remove the example that directly queries `_field_names`?

colings86 committed:
Addresses more review comments

colings86 committed:
implements existsQuery explicitly on every mapper

colings86 commented:
@jpountz I pushed another commit. This addresses your remaining comments including removing the implementation of `existsQuery` from `MappedFieldType` and instead requires (and implements) the method on the subclasses

jpountz commented on core/src/main/java/org/elasticsearch/index/mapper/CompletionFieldMapper.java:
```diff
@@ -257,6 +262,22 @@ public static synchronized PostingsFormat postingsFormat() {
 
+        @Override
+        public Query existsQuery(QueryShardContext context) {
+            if (hasDocValues()) {
```
this will always be false?

jpountz commented on core/src/main/java/org/elasticsearch/index/mapper/GeoShapeFieldMapper.java:
```diff
@@ -407,6 +416,22 @@ public PrefixTreeStrategy resolveStrategy(String strategyName) {
 
         @Override
+        public Query existsQuery(QueryShardContext context) {
+            if (hasDocValues()) {
```
always false?

jpountz commented on core/src/main/java/org/elasticsearch/index/mapper/RoutingFieldMapper.java:
```diff
@@ -121,6 +125,22 @@ public MappedFieldType clone() {
+
+        @Override
+        public Query existsQuery(QueryShardContext context) {
+            if (hasDocValues()) {
```
always false?

jpountz reviewed (approved):
It looks good to me overall! I have two questions left:
 - should we document the breaking change about term queries on `_field_names` or maybe even try to maintain backward compatibility? In any cases we should at least remove the term query example in the docs?
 - Doc value iterators are only good for sure if they were written with Lucene 7. So I'm wondering that we might still want to write field names entries if the index was created before 6.0 as well as only use DocValuesFieldExistsQuery if the index was created on or after 6.0?

colings86 commented:
> should we document the breaking change about term queries on _field_names or maybe even try to maintain backward compatibility? In any cases we should at least remove the term query example in the docs?

Apart from going back to always writing all field names to the `_field_names` field the only way I see to maintain backward compatibility for term queries on `_field_names` is to have `FieldNamesFieldMapper.termQuery()` retrieve the `MappedFieldType` for the field being queried and call `existsQuery()` on it. At the moment the implementations of `existsQuery`call `FieldNamesFieldMapper.termQuery()` so if we want to do this we would need to have the `existsQuery()` implementations instead create a `TermQuery` directly otherwise we would end up with a circular call chain. Does this sound ok to you?

> Doc value iterators are only good for sure if they were written with Lucene 7. So I'm wondering that we might still want to write field names entries if the index was created before 6.0 as well as only use DocValuesFieldExistsQuery if the index was created on or after 6.0?

Ok so this would mean checking the indexCreatedVersion at both index time and query time to determine how to index/query the data? It would also mean the index-time performance improvements would only be visible for indexes created after 6.0?

jpountz commented:
> Does this sound ok to you?

I think it is fine. It would also be symmetric since we add StringField instances manually at index time rather than calling FieldNamesFieldMapper.parse?

> Ok so this would mean checking the indexCreatedVersion at both index time and query time to determine how to index/query the data? It would also mean the index-time performance improvements would only be visible for indexes created after 6.0?

Correct.

colings86 committed:
Reinstates ability to perform term query on `_field_names`

nik9000 commented on core/src/main/java/org/elasticsearch/index/mapper/FieldNamesFieldMapper.java:
```diff
@@ -179,11 +181,20 @@ public boolean isEnabled() {
         }
 
         @Override
```
Should you override `parseCreateField` to throw an IllegalStateException or something? I don't think we use it any more.

colings86 replied:
We didn't use it until I pushed the latest commits which put the BWC code in to make sure `_field_names` is still used for indexes created before 6.1. I will make sure I change `parseCreateField` to throw an exception when I remove the BWC code though. Thanks for pointing this out.

colings86 committed:
Added bwc depending on index created version

colings86 commented on core/src/main/java/org/elasticsearch/index/mapper/BinaryFieldMapper.java:
```diff
@@ -165,6 +178,8 @@ protected void parseCreateField(ParseContext context, List<IndexableField> field
                 field.add(value);
             }
+        } else {
+            createFieldNamesField(context, fields);
```
Good point, will add

colings86 committed:
Review Comments

colings86 commented:
@nik9000 @jpountz thanks for the reviews so far. I've pushed some more commits, would you be able to take another look?

colings86 committed:
Skips tests that are not supported in 6.1.0

These values will need to be changed after backporting this PR to 6.x

jpountz commented on core/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java:
```diff
@@ -285,6 +287,16 @@ public Mapper parse(ParseContext context) throws IOException {
+                .metadataMapper(FieldNamesFieldMapper.class).fieldType();
+        if (fieldNamesFieldType != null && fieldNamesFieldType.isEnabled()) {
+            for (String fieldName : FieldNamesFieldMapper.extractFieldNames(fieldType().name())) {
+                    fields.add(new Field(FieldNamesFieldMapper.NAME, fieldName, fieldNamesFieldType));
```
indentation

jpountz reviewed (approved):
LGTM

Bek commented:
This is triggering a few thousand `DocValuesFieldExistsQuery` for us when using `exists` for a field type `object` because objects under the field have even more deeply nested documents. Queries taking a second before are taking around 30-40 seconds. Profiling is indicating that `"exists": "user"` causes the `DocValuesFieldExistsQuery` to be called on each and all subfields as if `"exists": "user.*"` is intended. `"exists": "user.id"` returns without performance degradation. Is this the expected behavior?
